pkg/pillar/cmd/domainmgr/containerd.go

```diff
@@ -0,0 +1,199 @@
+package domainmgr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/utils"
+	uuid "github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// containerd socket
+	ctrdSocket = "/run/containerd/containerd.sock"
+	// CtrdServicesNamespace containerd namespace for running containers
+	CtrdServicesNamespace = "eve-user-apps"
+
+	// root path to all containers
+	containersRoot = "/persist/rkt/pods/prepared"
+	// relative path to rootfs for an individual container
+	containerRootfsPath = "stage1/rootfs/opt/stage2/runx/"
+)
+
+// KeyValue a key-value pair
+type KeyValue struct {
+	Name  string
+	Value string
+}
+
+// getContainerPath return the path to the root of the container. This is *not*
+// necessarily the rootfs, which may be a layer below
+func getContainerPath(id string) string {
+	return path.Join(containersRoot, id)
+}
+
+// getContainerRootfs return the path to the root of the container filesystem
+func getContainerRootfs(id string) string {
+	return path.Join(getContainerPath(id), containerRootfsPath)
+}
+
+// containerdLoadImageTar load an image tar into the containerd content store
+func containerdLoadImageTar(filename string) (map[string]string, error) {
+	// load the content into the containerd content store
+	ctr, err := containerd.New(ctrdSocket)
+	if err != nil {
+		log.Errorf("could not connect to containerd socket at %s: %+s", ctrdSocket, err.Error())
+		return nil, err
+	}
+	defer ctr.Close()
+
+	ctrdCtx := namespaces.WithNamespace(context.Background(), CtrdServicesNamespace)
+
+	tarReader, err := os.Open(filename)
+	if err != nil {
+		log.Errorf("could not open tar file for reading at %s: %+s", filename, err.Error())
+		return nil, err
+	}
+
+	imgs, err := ctr.Import(ctrdCtx, tarReader)
```

Comment:
Hey @deitch given the troubles we used to have with tarReader with rkt -- can you please test this with some humongous container and see that the memory consuption stays decent?

Reply:
Happy to. The memory issues there, though, weren't around tarReader, but around issues related to in-memory compression, a few other side issues. I still will check, of course.

Reply:
Actually, now that I reread it, we aren't actually using tarReader per se. I just called the var that, but it is just a straight open file handle. It is possible that ctr does something odd in the end, so still will check.

Reply:
Ran a test with a 5+GB image, checked the memory usage for both the calling process and containerd.

* The calling process had a <1MB memory bump when it started running, and that was it. Didn't change over the entire time it was loading into containerd.
* containerd had no noticeable difference